docker/stan-wasm-server/logic/compilation.py

```diff
@@ -0,0 +1,98 @@
+import tempfile
+import subprocess
+from hashlib import sha1
+from shutil import copy2, rmtree
+from pathlib import Path
+
+from .definitions import CompilationStatus
+from .locking import get_nonce, acquire_compilation_lock, release_compilation_lock
+
+
+COMPILATION_TIMEOUT = 60 * 5
+COMPILATION_OUTPUTS = ['main.js', 'main.wasm']
+
+
+def _compute_stan_program_hash(program_file: Path):
+    stan_program = program_file.read_text()
+    # TODO: replace stan_program with a canonical form
+    return sha1(stan_program.encode()).hexdigest()
+
+
+def make_canonical_model_dir(job_id: str, tinystan_dir: str):
+    stan_code_file_name = Path(job_id, 'main.stan')
+    stan_program_hash = _compute_stan_program_hash(stan_code_file_name)
+    model_dir = Path(tinystan_dir, 'test_models', stan_program_hash)
+    model_dir.mkdir(exist_ok=True)
+    return model_dir
+
+
+def copy_compilation_outputs(*, model_dir: Path, job_dir: Path):
+    todos = []
+    for f in COMPILATION_OUTPUTS:
+        src = Path(model_dir, f)
+        dest = Path(job_dir, f)
+        if not src.exists():
+            raise FileNotFoundError(f"File {src} does not exist.")
+        todos.append((src, dest))
+    for (src, dest) in todos:
+        copy2(src, dest)
+
+
+def try_compile_stan_program(*, job_dir: Path, model_dir: Path, tinystan_dir: str, preserve_on_fail = False) -> tuple[CompilationStatus, str]:
+    try:
+        nonce = get_nonce()
+        if acquire_compilation_lock(model_dir, nonce):
+            compile_model_if_uncompiled(job_dir=job_dir, model_dir=model_dir, tinystan_dir=tinystan_dir, preserve_on_fail=preserve_on_fail)
+            copy_compilation_outputs(model_dir=model_dir, job_dir=job_dir)
+            return (CompilationStatus.COMPLETED, '')
+        # NOTE: Could also include job ID in lockfile and report that here
+        return (CompilationStatus.RUNNING, '')
+    except Exception as e:
+        return (CompilationStatus.FAILED, str(e))
+    finally:
+        release_compilation_lock(model_dir, nonce)
```

Comment:
Currently, `compile_model_if_uncompiled` blows away the model directory on timeout/error, which then causes this to raise an exception about not being able to find running.txt

Reply:
Yep, we were catching the wrong error (`FileNotFoundError` is quite different from `FileExistsError` as it turns out!)